File: NCoreUtils.AspNetCore.Rest/DefaultQueryFilter.fs
namespace NCoreUtils.AspNetCore.Rest

open System
open System.Linq
open System.Linq.Expressions
open Microsoft.Extensions.DependencyInjection
open NCoreUtils.Data

/// <summary>
/// Provides default query filtering implementation. Uses <see cref="NCoreUtils.Data.IDataQueryExpressionBuilder" /> for
/// processing rest query conditions.
/// </summary>
type DefaultQueryFilter<'a> =
  val private serviceProvider : IServiceProvider

  /// <summary>
  /// Initializes new instance from the specified parameters.
  /// </summary>
  /// <param name="serviceProvider">Service provider.</param>
  new (serviceProvider) = { serviceProvider = serviceProvider }

  abstract FilterQuery : queryable:IQueryable<'a> * restQuery:RestQuery -> IQueryable<'a>

  /// <summary>
  /// Filters queryable using provided REST query.
  /// </summary>
  /// <param name="queryable">Source queryable.</param>
  /// <param name="restQuery">REST query to apply.</param>
  /// <returns>Filtered queryable.</returns>
  default this.FilterQuery (queryable : IQueryable<'a>, restQuery : RestQuery) =
    match String.IsNullOrEmpty restQuery.Filter with
    | true -> queryable
    | _ ->
      let expressionBuilder = this.serviceProvider.GetRequiredService<IDataQueryExpressionBuilder> ()
      let predicate = expressionBuilder.BuildExpression<'a>(restQuery.Filter) :?> Expression<Func<'a, bool>>
      Queryable.Where (queryable, predicate)

  interface IRestQueryFilter<'a> with
    member this.FilterQuery (queryable, restQuery) = this.FilterQuery (queryable, restQuery)

File: NCoreUtils.AspNetCore.Rest/Rest/DefaultQueryFilter.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Linq.Expressions;
using NCoreUtils.Data.Protocol;

namespace NCoreUtils.AspNetCore.Rest;

public class DefaultQueryFilter<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T>(IDataQueryExpressionBuilder? queryExpressionBuilder = null) : IRestQueryFilter<T>
{
    private readonly IDataQueryExpressionBuilder? _queryExpressionBuilder = queryExpressionBuilder;

    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Should be handled by the provider.")]
    public IQueryable<T> ApplyFilters(IQueryable<T> source, RestQuery restQuery)
    {
        if (string.IsNullOrWhiteSpace(restQuery.Filter))
        {
            return source;
        }
        if (_queryExpressionBuilder is null)
        {
            throw new InvalidOperationException("Default rest query filter requires NCoreUtils data query services in order to parse filters.");
        }
        Expression<Func<T, bool>> predicate;
        try
        {
            var expression = _queryExpressionBuilder.BuildExpression(typeof(T), restQuery.Filter);
            if (expression.Body.TryExtractConstant(out var cbox))
            {
                var cbool = (bool)Convert.ChangeType(cbox, typeof(bool))!;
                predicate = Expression.Lambda<Func<T, bool>>(QueryableExtensions.BoxConstant(cbool), expression.Parameters);
            }
            else
            {
                predicate = (Expression<Func<T, bool>>)expression;
            }
        }
        catch (Exception exn)
        {
            throw new BadRequestException($"Invalid filter has been specified.", exn);
        }
        return source.Where(predicate);
    }
}